bonsai/UseCases/BudgetGateway.swift

```diff
@@ -0,0 +1,83 @@
+//
+//  BudgetGateway.swift
+//  bonsai
+//
+//  Created by hoang on 21.12.2021.
+//
+
+import Foundation
+import CoreData
+
+public protocol BudgetProtocol {
+   var budgetID: String { get }
+   var name: String { get }
+   var amount: Double { get }
+   var period: Int64 { get } // days ??
+}
+
+public final class BudgetGateway {
+
+   let mainContext: NSManagedObjectContext
+
+   init(mainContext: NSManagedObjectContext = DataController.sharedInstance.mainContext) {
+      self.mainContext = mainContext
+   }
+
+   @discardableResult
+   func create(name: String, amount: Double, period: Int64) throws -> Budget {
+      let budget = Budget(
+         context: mainContext,
+         name: name,
+         amount: amount,
+         period: period)
+      try mainContext.save()
+      return budget
+   }
+
+   func update(budget: Budget) throws {
+      try mainContext.save()
+   }
+
+   func delete(budget: Budget) throws {
+      mainContext.delete(budget)
+      try mainContext.save()
+   }
+
+   func getBudget(name: String) throws -> BudgetProtocol {
+      let request: NSFetchRequest<Budget> = Budget.fetchRequest()
+      request.fetchLimit = 1
+      request.predicate = NSPredicate(format: "name == %@", name)
+      guard let budget = try mainContext.fetch(request).first else {
+         throw BudgetDoesntExist()
+      }
+      return budget
+   }
+
+   func getAllBudgets() throws -> [BudgetProtocol] {
+      let request: NSFetchRequest<Budget> = Budget.fetchRequest()
+      let budgets = try mainContext.fetch(request)
+      return budgets
+   }
+
+   func spentCapacity(budgetName: String) throws -> Double {
+      let budget = try getBudget(name: budgetName)
+      return (budget.amount) / Double(budget.period).roundTo(places: 1)
+   }
+
+   func spentCapacityLessThanAvailable(budget: Budget) throws -> Double {
+      let capacity = try spentCapacity(budgetName: budget.name)
```

Comment:
Usage of Decimal calculations is must have if we calculate something related to money